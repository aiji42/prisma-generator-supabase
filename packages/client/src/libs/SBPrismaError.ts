export class SBPrismaError extends Error {
  constructor(e?: string) {
    super(e)
    this.name = new.target.name

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor)
    }
  }
}

export class SBPrismaAPIError extends SBPrismaError {
  constructor(e: { message: string; [k: string]: string | null }) {
    console.error(e)
    super(e.message)
  }
}
