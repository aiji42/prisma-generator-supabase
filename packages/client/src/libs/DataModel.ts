import { DMMF } from '@prisma/generator-helper'
import { SBPrismaError } from './SBPrismaError'

export class DataModel {
  cache: Map<string, Model> = new Map()

  constructor(public dataModel: DMMF.Datamodel) {}

  public model(key: string) {
    const cache = this.cache.get(key)
    if (cache) return cache

    const model = this.dataModel.models.find(({ name }) => name === key)
    if (!model)
      throw new SBPrismaError(
        `Model (${key}) does not exist in dmmf.datamodel.models.`,
      )

    const m = new Model(model)
    this.cache.set(key, m)
    return m
  }
}

class Model {
  cache: Map<string, DMMF.Field> = new Map()

  constructor(public model: DMMF.Model) {}

  get table() {
    return this.model.dbName ?? this.model.name
  }

  get fields() {
    return this.model.fields
  }

  public field(key: string) {
    const cache = this.cache.get(key)
    if (cache) return cache

    const field = this.model.fields.find(({ name }) => name === key)
    if (!field)
      throw new SBPrismaError(
        `Field (${key}) does not exist in dmmf.datamodel.models.[].fields`,
      )

    this.cache.set(key, field)
    return field
  }
}
