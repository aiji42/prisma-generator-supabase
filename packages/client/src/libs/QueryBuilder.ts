import { Args, ArgsFlatten, Select, SelectFlatten } from '../types'
import { DMMF } from '@prisma/generator-helper'
import { DataModel } from './DataModel'
import { makeSelect } from '../helpers/makeSelect'
import { makeWhere } from '../helpers/makeWhere'
import { makeOrder } from '../helpers/makeOrder'

export class QueryBuilder {
  doc: DataModel
  args: Args = {}
  argsFlattens: ArgsFlatten[] = []
  rootModel: string

  constructor(dataModel: DMMF.Document['datamodel']) {
    this.doc = new DataModel(dataModel)
  }

  setup(args: Args, model: string) {
    this.args = args
    this.rootModel = model
    this.argsFlattens = flat(this.rootModel, this.args, '', this.doc)
    return this
  }

  public toString() {
    return this.query.toString()
  }

  get query() {
    const params = new URLSearchParams()
    params.append('select', this.select)
    this.wheres.forEach(([prefix, where]) => {
      params.append(prefix ? `${prefix}.and` : 'and', `(${where})`)
    })
    this.orders.forEach(([prefix, order]) => {
      params.append(prefix ? `${prefix}.order` : 'order', order)
    })
    this.limits.forEach(([prefix, limit]) => {
      params.append(prefix ? `${prefix}.limit` : 'limit', limit)
    })
    this.offset.forEach(([prefix, offset]) => {
      params.append(prefix ? `${prefix}.offset` : 'offset', offset)
    })

    return params
  }

  get select(): string {
    return makeSelect(this.args, this.rootModel, this.doc)
  }

  get wheres(): [string, string][] {
    return this.argsFlattens.reduce<[string, string][]>(
      (res, { model, prefix, ...args }) => {
        const where = makeWhere(args, model, this.doc)
        return where ? [...res, [prefix, where]] : res
      },
      [],
    )
  }
  get orders(): [string, string][] {
    return this.argsFlattens.reduce<[string, string][]>(
      (res, { prefix, ...args }) => {
        const order = makeOrder(args)
        return order ? [...res, [prefix, order]] : res
      },
      [],
    )
  }
  get limits(): [string, string][] {
    return this.argsFlattens.reduce<[string, string][]>(
      (res, { prefix, take }) => {
        return typeof take === 'number' ? [...res, [prefix, String(take)]] : res
      },
      [],
    )
  }
  get offset(): [string, string][] {
    return this.argsFlattens.reduce<[string, string][]>(
      (res, { prefix, skip }) => {
        return typeof skip === 'number' ? [...res, [prefix, String(skip)]] : res
      },
      [],
    )
  }
}

const flat = (
  model: string,
  args: Args,
  prefix: string,
  doc: DataModel,
): ArgsFlatten[] => {
  const { select = {}, data, skipDuplicates, ...rest } = args
  if (select._count) {
    return flat(model, { select: select._count.select, ...rest }, prefix, doc)
  }
  const [rootSelects, children] = Object.entries(select).reduce<
    [SelectFlatten, Select]
  >(
    (res, [columnOrModel, fields]) => {
      const [root, children] = res
      if (
        columnOrModel === '_all' ||
        doc.model(model).field(columnOrModel).kind !== 'object'
      )
        return [{ ...root, [columnOrModel]: fields as boolean }, children]
      return [root, { ...children, [columnOrModel]: fields }]
    },
    [{}, {}],
  )

  return [
    {
      select: rootSelects,
      ...rest,
      prefix,
      model,
    },
    ...Object.entries(children).flatMap(([alias, child]) =>
      flat(
        doc.model(model).field(alias).type,
        child as Args,
        prefix ? `${prefix}.${alias}` : alias,
        doc,
      ),
    ),
  ]
}
