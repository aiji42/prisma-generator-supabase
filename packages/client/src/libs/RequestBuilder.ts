import { QueryBuilder } from './QueryBuilder'
import { DataModel } from './DataModel'
import { DMMF } from '@prisma/generator-helper'
import { Args } from '../types'
import { v4 as uuid } from 'uuid'
import crossFetch from 'cross-fetch'
import { SBPrismaAPIError, SBPrismaError } from './SBPrismaError'

type Fetch = typeof fetch

export class RequestBuilder {
  queryBuilder: QueryBuilder
  doc: DataModel

  constructor(
    dataModel: DMMF.Datamodel,
    private _endpoint: string,
    private _apikey: string,
  ) {
    this.queryBuilder = new QueryBuilder(dataModel)
    this.doc = this.queryBuilder.doc
  }

  public build(args: Args, model: string, action: string) {
    const url = new URL(`${this._endpoint}`)
    url.pathname = `/rest/v1/${this.doc.model(model).table}`
    url.search = String(this.queryBuilder.setup(args, model))

    const headers = {
      apikey: this._apikey,
      Authorization: `Bearer ${this._apikey}`,
      'Content-Type': 'application/json',
    }

    if (['findFirst', 'findUnique', 'findMany'].includes(action)) {
      return findRequest(url, headers)
    }
    if (['delete', 'deleteMany'].includes(action)) {
      return deleteRequest(url, headers)
    }
    if (['update', 'updateMany'].includes(action)) {
      return updateRequest(url, headers, args, model, this.doc)
    }
    if (['create', 'createMany'].includes(action)) {
      return createRequest(url, headers, args, model, this.doc)
    }

    throw new SBPrismaError(`Action ${action} is not yet supported.`)
  }

  public async execute(args: Args, model: string, action: string) {
    const req = this.build(args, model, action)
    const res = await (typeof fetch === 'undefined' ? crossFetch : fetch)(
      ...req,
    )
    return await afterHandler(res, action)
  }

  set apikey(key: string) {
    this._apikey = key
  }
}

const autoUpdatedAt = (model: string, data: Args['data'], doc: DataModel) => {
  if (!data) return data
  const now = new Date()
  const filler = doc
    .model(model)
    .fields.reduce<Record<string, Date>>((res, field) => {
      if (
        !(
          field.isUpdatedAt &&
          typeof field.default === 'object' &&
          field.default.name === 'now'
        )
      )
        return res
      return { ...res, [field.name]: now }
    }, {})

  if (Array.isArray(data)) return data.map((d) => ({ ...filler, ...d }))
  return { ...filler, ...data }
}

const autoId = (model: string, data: Args['data'], doc: DataModel) => {
  if (!data) return data
  const columns = doc.model(model).fields.reduce<string[]>((res, field) => {
    if (typeof field.default !== 'object') return res
    if (field.default.name === 'uuid') return [...res, field.name]
    if (field.default.name === 'cuid')
      throw new SBPrismaError(
        'Not supported `@default(cuid())`. Please use `@default(uuid())`',
      )
    return res
  }, [])
  if (columns.length < 1) return data

  if (Array.isArray(data))
    return data.map((d) => ({
      ...Object.fromEntries(columns.map((col) => [col, uuid()])),
      ...d,
    }))

  return { ...Object.fromEntries(columns.map((col) => [col, uuid()])), ...data }
}

const findRequest = (url: URL, headers: HeadersInit): Parameters<Fetch> => {
  return [
    url.toString(),
    {
      method: 'GET',
      headers,
    },
  ]
}

const deleteRequest = (url: URL, headers: HeadersInit): Parameters<Fetch> => {
  return [
    url.toString(),
    {
      method: 'DELETE',
      headers: { ...headers, Prefer: 'return=representation' },
    },
  ]
}

const updateRequest = (
  url: URL,
  headers: HeadersInit,
  { data }: Args,
  model: string,
  doc: DataModel,
): Parameters<Fetch> => {
  return [
    url.toString(),
    {
      method: 'PATCH',
      headers: { ...headers, Prefer: 'return=representation' },
      body: JSON.stringify(autoUpdatedAt(model, data, doc)),
    },
  ]
}

const createRequest = (
  url: URL,
  headers: HeadersInit,
  { data, skipDuplicates }: Args,
  model: string,
  doc: DataModel,
): Parameters<Fetch> => {
  return [
    url.toString(),
    {
      method: 'POST',
      headers: {
        ...headers,
        Prefer: `return=representation${
          skipDuplicates ? ',resolution=ignore-duplicates' : ''
        }`,
      },

      body: JSON.stringify(autoId(model, data, doc)),
    },
  ]
}

const singly = (data: unknown[]): unknown | undefined => {
  return data[0]
}

const afterHandler = async (res: Response, action: string) => {
  if (!res.ok) throw new SBPrismaAPIError(await res.json())

  if (
    ['findFirst', 'findUnique', 'delete', 'update', 'create'].includes(action)
  ) {
    return singly(await res.json())
  }
  if (action === 'findMany') {
    return await res.json()
  }
  if (['deleteMany', 'updateMany', 'createMany'].includes(action)) {
    return { count: (await res.json()).length }
  }

  throw new SBPrismaError(`Action ${action} is not yet supported.`)
}
