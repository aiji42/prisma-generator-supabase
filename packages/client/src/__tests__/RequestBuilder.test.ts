import { getSampleDMMF } from './__fixtures__/getSampleSchema'
import { RequestBuilder } from '../libs/RequestBuilder'
import { Where } from '../types'
import crossFetch from 'cross-fetch'

jest.mock('cross-fetch', () => {
  //Mock the default export
  return {
    __esModule: true,
    default: jest.fn(),
  }
})

jest.mock('uuid', () => ({
  v4: () => {
    return 'this is uuid mocked value'
  },
}))

beforeEach(() => {
  const fixed = new Date('2022-05-10T00:00:00')
  jest.useFakeTimers().setSystemTime(fixed.getTime())
})

afterEach(() => {
  jest.useRealTimers()
  jest.resetAllMocks()
})

test('RequestBuilder; build - findMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  const req = queryBuilder.build(
    {
      select: { name: true },
      where: {
        name: { startsWith: 'F' },
        OR: [{ id: { gt: 10, lt: 20 } }, { id: { gte: 110, lte: 120 } }],
      } as Where,
      take: 10,
      skip: 100,
    },
    'User',
    'findMany',
  )

  expect(req).toMatchSnapshot()
})

test('RequestBuilder; build - count', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  const req = queryBuilder.build(
    {
      select: { name: true, _all: true },
      where: {
        NOT: [
          { name: { startsWith: 'z' } },
          { name: { not: { equals: 'a' } } },
        ],
      } as Where,
    },
    'Team',
    'count',
  )

  expect(req).toMatchSnapshot()
})

test('RequestBuilder; build - creatMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  const req = queryBuilder.build(
    {
      data: [{ name: 'Foo' }, { name: 'Baa' }],
    },
    'User',
    'createMany',
  )

  expect(req).toMatchSnapshot()
})

test('RequestBuilder; build - creatMany with skipDuplicates', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  const req = queryBuilder.build(
    {
      data: [{ name: 'Foo' }, { name: 'Baa' }],
      skipDuplicates: true,
    },
    'User',
    'createMany',
  )

  expect(req).toMatchSnapshot()
})

test('RequestBuilder; build - updateMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  const req = queryBuilder.build(
    {
      data: { name: 'Foo' },
      where: {
        name: { not: { startsWith: 'Foo' } },
      },
    },
    'User',
    'updateMany',
  )

  expect(req).toMatchSnapshot()
})

test('RequestBuilder; build - deleteMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  const req = queryBuilder.build(
    {
      where: {
        name: { not: { startsWith: 'Foo' } },
      },
    },
    'User',
    'deleteMany',
  )

  expect(req).toMatchSnapshot()
})

test('RequestBuilder; execute - findMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  ;(crossFetch as jest.Mock).mockResolvedValue({
    status: 200,
    ok: true,
    json: () => [
      {
        id: 'aaaa',
        name: 'Foo',
        updatedAt: new Date().toISOString(),
      },
      {
        id: 'bbbb',
        name: 'Boo',
        updatedAt: new Date().toISOString(),
      },
    ],
  })

  const res = await queryBuilder.execute(
    { select: { id: true, name: true, updatedAt: true } },
    'User',
    'findMany',
  )

  expect(res).toMatchSnapshot()
})

test('RequestBuilder; execute - findFirst', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  ;(crossFetch as jest.Mock).mockResolvedValue({
    status: 200,
    ok: true,
    json: () => [
      {
        id: 'aaaa',
        name: 'Foo',
        updatedAt: new Date().toISOString(),
      },
      {
        id: 'bbbb',
        name: 'Boo',
        updatedAt: new Date().toISOString(),
      },
    ],
  })

  const res = await queryBuilder.execute(
    { select: { id: true, name: true, updatedAt: true } },
    'User',
    'findFirst',
  )

  expect(res).toMatchSnapshot()
})

test('RequestBuilder; execute - deleteMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  ;(crossFetch as jest.Mock).mockResolvedValue({
    status: 200,
    ok: true,
    json: () => [
      {
        id: 'aaaa',
        name: 'Foo',
        updatedAt: new Date().toISOString(),
      },
      {
        id: 'bbbb',
        name: 'Boo',
        updatedAt: new Date().toISOString(),
      },
    ],
  })

  const res = await queryBuilder.execute({}, 'User', 'deleteMany')

  expect(res).toMatchSnapshot()
})

test('RequestBuilder; execute - updateMany', async () => {
  const queryBuilder = new RequestBuilder(
    (await getSampleDMMF()).datamodel,
    'https://example.com',
    'apikey',
  )
  ;(crossFetch as jest.Mock).mockResolvedValue({
    status: 200,
    ok: true,
    json: () => [
      {
        id: 'aaaa',
        name: 'Foo',
        updatedAt: new Date().toISOString(),
      },
      {
        id: 'bbbb',
        name: 'Boo',
        updatedAt: new Date().toISOString(),
      },
    ],
  })

  const res = await queryBuilder.execute({}, 'User', 'updateMany')

  expect(res).toMatchSnapshot()
})
