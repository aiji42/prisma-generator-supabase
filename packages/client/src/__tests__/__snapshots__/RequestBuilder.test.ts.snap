// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestBuilder; build - creatMany 1`] = `
Array [
  "https://example.com/rest/v1/User?select=*",
  Object {
    "body": "[{\\"id\\":\\"this is uuid mocked value\\",\\"name\\":\\"Foo\\"},{\\"id\\":\\"this is uuid mocked value\\",\\"name\\":\\"Baa\\"}]",
    "headers": Object {
      "Authorization": "Bearer apikey",
      "Content-Type": "application/json",
      "Prefer": "return=representation",
      "apikey": "apikey",
    },
    "method": "POST",
  },
]
`;

exports[`RequestBuilder; build - creatMany with skipDuplicates 1`] = `
Array [
  "https://example.com/rest/v1/User?select=*",
  Object {
    "body": "[{\\"id\\":\\"this is uuid mocked value\\",\\"name\\":\\"Foo\\"},{\\"id\\":\\"this is uuid mocked value\\",\\"name\\":\\"Baa\\"}]",
    "headers": Object {
      "Authorization": "Bearer apikey",
      "Content-Type": "application/json",
      "Prefer": "return=representation,resolution=ignore-duplicates",
      "apikey": "apikey",
    },
    "method": "POST",
  },
]
`;

exports[`RequestBuilder; build - deleteMany 1`] = `
Array [
  "https://example.com/rest/v1/User?select=*&and=%28name.not.like.Foo*%29",
  Object {
    "headers": Object {
      "Authorization": "Bearer apikey",
      "Content-Type": "application/json",
      "Prefer": "return=representation",
      "apikey": "apikey",
    },
    "method": "DELETE",
  },
]
`;

exports[`RequestBuilder; build - findMany 1`] = `
Array [
  "https://example.com/rest/v1/User?select=name&and=%28or%28id.lt.20%2Cid.gt.10%2Cid.lte.120%2Cid.gte.110%29%2Cname.like.F*%29&limit=10&offset=100",
  Object {
    "headers": Object {
      "Authorization": "Bearer apikey",
      "Content-Type": "application/json",
      "apikey": "apikey",
    },
    "method": "GET",
  },
]
`;

exports[`RequestBuilder; build - updateMany 1`] = `
Array [
  "https://example.com/rest/v1/User?select=*&and=%28name.not.like.Foo*%29",
  Object {
    "body": "{\\"updatedAt\\":\\"2022-05-10T00:00:00.000Z\\",\\"name\\":\\"Foo\\"}",
    "headers": Object {
      "Authorization": "Bearer apikey",
      "Content-Type": "application/json",
      "Prefer": "return=representation",
      "apikey": "apikey",
    },
    "method": "PATCH",
  },
]
`;

exports[`RequestBuilder; execute - deleteMany 1`] = `
Object {
  "count": 2,
}
`;

exports[`RequestBuilder; execute - findFirst 1`] = `
Object {
  "id": "aaaa",
  "name": "Foo",
  "updatedAt": "2022-05-10T00:00:00.000Z",
}
`;

exports[`RequestBuilder; execute - findMany 1`] = `
Array [
  Object {
    "id": "aaaa",
    "name": "Foo",
    "updatedAt": "2022-05-10T00:00:00.000Z",
  },
  Object {
    "id": "bbbb",
    "name": "Boo",
    "updatedAt": "2022-05-10T00:00:00.000Z",
  },
]
`;

exports[`RequestBuilder; execute - updateMany 1`] = `
Object {
  "count": 2,
}
`;
